src/back/spv/writer.rs

```diff
@@ -1115,59 +1115,56 @@ impl Writer {
                 };
 
                 let (instruction, lookup_ty) = match op {
-                    crate::BinaryOperator::Multiply => match *left_ty_inner {
-                        crate::TypeInner::Vector { .. } => match *right_ty_inner {
-                            crate::TypeInner::Scalar { .. } => (
-                                super::instructions::instruction_vector_times_scalar(
-                                    left_result_type_id,
-                                    id,
-                                    left_id,
-                                    right_id,
-                                ),
-                                left_lookup_ty,
+                    crate::BinaryOperator::Multiply => match (*left_ty_inner, *right_ty_inner) {
```

Comment:
Why can I match on reference with a single value, but not with two? Has it something to do that ( ) is a tuple like structure where Rust stores the values in?

Do you have any idea how to fix this, @kvark ? Maybe your proposed solution like discussed at Matrix, e.g.:
```rust
fn dimension(ty: &crate::TypeInner) -> u8 {
  match *ty {
    Ty::Scalar { .. } => 0,
    Ty::Vector { .. } => 1,
    Ty::Matrix { .. } => 2,
    _ => unreachable!(),
  }
}
```

Reply:
the main concept to understand is that pattern matching on something doesn't take it by value, i.e. `match x` doesn't require moving of `x` anywhere. So `match *x` works fine if you just have `x: &Something`. However, as soon as you write `(*x, *y)`, Rust can't see through this expression, the smartness is limited to the immediate argument of `match`. So `(*x, *y)` has to move things out, and we have to resort to `(x, y)` (where these are references) instead.

Reply:
I think we should do with a helper `fn dimension(ty: &crate::TypeInner) -> Dimension` to work around that matching limitation